{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Yousuf's Method. The input is meant to be a list of equal-sized k-mers and the goal is figure out the total of number of \n",
    "## nucleotide counts at each posiiton and return that as a dictionary of lists (each list is a different nucleotide)\n",
    "def Count(Motifs):\n",
    "    #I am creating a dictionary of nucleotides which will contain a list of nt frequency at each position\n",
    "    count = {\"A\": [], \"C\": [], \"G\": [], \"T\": []}\n",
    "    #d represents the number of different sequences I want to find the motif in\n",
    "    d = len(Motifs)\n",
    "    #k represents the length of the kmer I'm looking at... since all of them are the same length I'm just looking at\n",
    "    #the first one\n",
    "    k = len(Motifs[0])\n",
    "    #I am first iterating through the positions of each motif, starting at position zero and ending with the final position k\n",
    "    for i in range(k):\n",
    "        A = 0\n",
    "        T = 0\n",
    "        C = 0\n",
    "        G = 0\n",
    "        #I am iterating through every string at position i and will add up how many nucleotides go to what then append the \n",
    "        #number to the list\n",
    "        for kmer_number,kmer in enumerate(Motifs):\n",
    "            if kmer[i] == \"A\":\n",
    "                A +=1\n",
    "            if kmer[i] == \"T\":\n",
    "                T +=1\n",
    "            if kmer[i] == \"C\":\n",
    "                C +=1\n",
    "            if kmer[i] == \"G\":\n",
    "                G+=1\n",
    "            if kmer_number == len(Motifs)-1:\n",
    "                count[\"A\"].append(A)\n",
    "                count[\"T\"].append(T)\n",
    "                count[\"C\"].append(C)\n",
    "                count[\"G\"].append(G)\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Using the hints/advice given to me by the program, I will design this second \n",
    "\n",
    "def Count2(Motifs):\n",
    "    #Making an empty dictionary\n",
    "    count = {}\n",
    "    #Creating the dictionary with the appropriate number of zeros in every position for every nucleotide already inputted\n",
    "    #k is th length of the motif. Each motif is the same length so we can just use the first one\n",
    "    #The list iterates through the 4 nucleotides of life, ACGT and creates an empty list [] with the key being the symbol\n",
    "    #Once an empty list is created with the key as the nucleotide symbol, a list of zeros is added\n",
    "    #The number of zeros corresponds to k\n",
    "    k = len(Motifs[0])\n",
    "    for symbol in \"ACGT\":\n",
    "        count[symbol] = []\n",
    "        for j in range(k):\n",
    "             count[symbol].append(0)\n",
    "    #t is the number of motifs we need t count\n",
    "    t = len(Motifs)\n",
    "    #iterating from 0 the the max number of sequences\n",
    "    for i in range(t):\n",
    "        #For each motif, we will iterate through the length\n",
    "        for j in range(k):\n",
    "            #We are going to read what the nucleotide is at the motif we're looking at --> i and what position --> j\n",
    "            symbol = Motifs[i][j]\n",
    "            #add +1 to the dictionary \n",
    "            count[symbol][j] += 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is a program that will normalize the nucleotide counts at each position to a decimal frequency \n",
    "def Profile(Motifs):\n",
    "    #number of motifs we have to count\n",
    "    t = len(Motifs)\n",
    "    #length of each motif..all of them are the same length so we just count the length of the first one\n",
    "    k = len(Motifs[0])\n",
    "    #getting our raw nucleotide counts from the count program\n",
    "    count = Count(Motifs)\n",
    "    #setting up a new dictionary using the technique from count2 where all the positions are 0's\n",
    "    normalized_count = {}\n",
    "    for symbol in \"ACGT\":\n",
    "        normalized_count[symbol] = []\n",
    "        for j in range(k):\n",
    "             normalized_count[symbol].append(0)\n",
    "    #iterating through all the positions starting from 0\n",
    "    for position in range(k):\n",
    "        total = 0\n",
    "        #count the total number of nucleotide counts at each position\n",
    "        for nt in \"ACGT\":\n",
    "            total += count[nt][position]\n",
    "        #divide the raw nucleotide total by the total number of all nucleotides\n",
    "        for nt in \"ACGT\":\n",
    "            normalized_count[nt][position] = float(count[nt][position])/total\n",
    "    return normalized_count\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is a program that will give me back a consensus sequence based on count\n",
    "def Consensus(Motifs):\n",
    "    consensus = \"\"\n",
    "    k = len(Motifs[0])\n",
    "    count = Count(Motifs)\n",
    "    for position in range(k):\n",
    "        max = 0\n",
    "        max_nt = ''\n",
    "        for nt in \"ACGT\":\n",
    "            if count[nt][position] > max:\n",
    "                max = count[nt][position]\n",
    "                max_nt = nt\n",
    "        consensus += max_nt\n",
    "    return consensus\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is a program that will score the consensus sequence based on how many nucleotides DO NOT match the consensus sequence\n",
    "def Score(Motifs):\n",
    "    consensus = Consensus(Motifs)\n",
    "    count = Count(Motifs)\n",
    "    k = len(Motifs[0])\n",
    "    score = 0\n",
    "    for position in range(k):\n",
    "        current_nt = consensus[position]\n",
    "        for symbol in \"ACGT\":\n",
    "            if symbol != current_nt:\n",
    "                score += count[symbol][position]\n",
    "    return score\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This program will calculate the probability of obtaining a k-mer given the output(probability matrix) from Profile \n",
    "#See Biology Meets Programming - Motif Counter if confused\n",
    "def Pr(Text, Profile):\n",
    "    k = len(Text)\n",
    "    probability = 1\n",
    "    matrix = Profile\n",
    "    for position in range(k):\n",
    "        probability*= Profile[Text[position]][position]\n",
    "    return probability\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This program will find the MOST likely k-mer given the: k-mer size, the profile matrix, and sequence we want to search through\n",
    "def ProfileMostProbablePattern(text, k, profile):\n",
    "    #Since this sequence is circular, we need to add a bit of the beginning of the sequence to end for long motifs\n",
    "    extendedGenome = text + text[0:len(text)//2]\n",
    "    #just initializing some variables\n",
    "    best_prob = -1\n",
    "    best_kmer = extendedGenome[0:k]\n",
    "    #So first I'll be iterating through each position\n",
    "    for i in range(len(text)-k+1):\n",
    "        #Checks probability of k-mer i to i+k\n",
    "        current_prob = Pr(extendedGenome[i:i+k], profile)\n",
    "        #if probability is better than previous one, replace.\n",
    "        if current_prob > best_prob:\n",
    "            best_prob = current_prob\n",
    "            best_kmer = extendedGenome[i:i+k]\n",
    "    return best_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This program will find the best k-mer given: list of sequence strings, size of k-mer, number of sequences to search through \n",
    "def GreedyMotifSearch(Dna, k, t):\n",
    "    #The best motif is set as the first k-mer of each sequence, so you have a list of t-number of k-mers\n",
    "    BestMotifs = []\n",
    "    for i in range(0, t):\n",
    "        BestMotifs.append(Dna[i][0:k])\n",
    "    #Now n is the length of the first sequence I have\n",
    "    n = len(Dna[0])\n",
    "    #i will be the starting position of where I will be drawing my k-mers from\n",
    "    for i in range(n-k+1):\n",
    "        #Empty list creation\n",
    "        Motifs = []\n",
    "        #I add a k-mer of appropriate length from the first sequence string to the list Motifs\n",
    "        Motifs.append(Dna[0][i:i+k])\n",
    "        #iterating from 0 to the number of sequences you have\n",
    "        for j in range(1, t):\n",
    "            #P is equal to the probability matrix generated by profile, which will count how similar nucleotides are and return\n",
    "            #it as a percentage of A's or T's at a position. This will only have up to t-number of k-mers in it\n",
    "            P = Profile(Motifs[0:j])\n",
    "            #Add the most similar k-mer from the following sequences 1,2,3...t to this motifs list\n",
    "            Motifs.append(ProfileMostProbablePattern(Dna[j], k, P))\n",
    "            #Now we will check to see if the score of our generated 5 k-mers is better than the previous one. Thus we end up\n",
    "            #eventually with the 5 best k-mers \n",
    "        if Score(Motifs) < Score(BestMotifs):\n",
    "            BestMotifs = Motifs\n",
    "    return BestMotifs\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "####The code below this is for revising the previous greedy motif method into a better one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is a revised method of counting, since the original method suffers from Cromwell's problem i.e. assigning something a \n",
    "#probability of zero is not the best way to create a set... everything should have a nonzero percent chance of happening,\n",
    "#no matter how unlikely\n",
    "def CountWithPseudocounts(Motifs):\n",
    "    #Making an empty dictionary\n",
    "    count = {}\n",
    "    #Creating the dictionary with the appropriate number of zeros in every position for every nucleotide already inputted\n",
    "    #k is th length of the motif. Each motif is the same length so we can just use the first one\n",
    "    #The list iterates through the 4 nucleotides of life, ACGT and creates an empty list [] with the key being the symbol\n",
    "    #Once an empty list is created with the key as the nucleotide symbol, a list of zeros is added\n",
    "    #The number of zeros corresponds to k\n",
    "    k = len(Motifs[0])\n",
    "    for symbol in \"ACGT\":\n",
    "        count[symbol] = []\n",
    "        for j in range(k):\n",
    "             count[symbol].append(1)\n",
    "    #t is the number of motifs we need t count\n",
    "    t = len(Motifs)\n",
    "    #iterating from 0 the the max number of sequences\n",
    "    for i in range(t):\n",
    "        #For each motif, we will iterate through the length\n",
    "        for j in range(k):\n",
    "            #We are going to read what the nucleotide is at the motif we're looking at --> i and what position --> j\n",
    "            symbol = Motifs[i][j]\n",
    "            #add +1 to the dictionary \n",
    "            count[symbol][j] += 1\n",
    "    return count    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Revised ProfileCount\n",
    "def ProfileWithPseudocounts(Motifs):\n",
    "    #number of motifs we have to count\n",
    "    t = len(Motifs)\n",
    "    #length of each motif..all of them are the same length so we just count the length of the first one\n",
    "    k = len(Motifs[0])\n",
    "    #getting our raw nucleotide counts from the count program\n",
    "    count = CountWithPseudocounts(Motifs)\n",
    "    #setting up a new dictionary using the technique from count2 where all the positions are 0's\n",
    "    normalized_count = {}\n",
    "    for symbol in \"ACGT\":\n",
    "        normalized_count[symbol] = []\n",
    "        for j in range(k):\n",
    "             normalized_count[symbol].append(0)\n",
    "    #iterating through all the positions starting from 0\n",
    "    for position in range(k):\n",
    "        total = 0\n",
    "        #count the total number of nucleotide counts at each position\n",
    "        for nt in \"ACGT\":\n",
    "            total += count[nt][position]\n",
    "        #divide the raw nucleotide total by the total number of all nucleotides\n",
    "        for nt in \"ACGT\":\n",
    "            normalized_count[nt][position] = float(count[nt][position])/total\n",
    "    return normalized_count\n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Updated GreedyCount \n",
    "def GreedyMotifSearchWithPseudocounts(Dna, k, t):\n",
    "\n",
    "    #The best motif is set as the first k-mer of each sequence, so you have a list of t-number of k-mers\n",
    "    BestMotifs = []\n",
    "    for i in range(0, t):\n",
    "        BestMotifs.append(Dna[i][0:k])\n",
    "    #Now n is the length of the first sequence I have\n",
    "    n = len(Dna[0])\n",
    "    #i will be the starting position of where I will be drawing my k-mers from\n",
    "    for i in range(n-k+1):\n",
    "        #Empty list creation\n",
    "        Motifs = []\n",
    "        #I add a k-mer of appropriate length from the first sequence string to the list Motifs\n",
    "        Motifs.append(Dna[0][i:i+k])\n",
    "        #iterating from 0 to the number of sequences you have\n",
    "        for j in range(1, t):\n",
    "            #P is equal to the probability matrix generated by profile, which will count how similar nucleotides are and return\n",
    "            #it as a percentage of A's or T's at a position. This will only have up to t-number of k-mers in it\n",
    "            P = ProfileWithPseudocounts(Motifs[0:j])\n",
    "            #Add the most similar k-mer from the following sequences 1,2,3...t to this motifs list\n",
    "            Motifs.append(ProfileMostProbablePattern(Dna[j], k, P))\n",
    "            #Now we will check to see if the score of our generated 5 k-mers is better than the previous one. Thus we end up\n",
    "            #eventually with the 5 best k-mers \n",
    "        if Score(Motifs) < Score(BestMotifs):\n",
    "            BestMotifs = Motifs\n",
    "    return BestMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is a new way to find motifs in sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Find the best k-mer from each sequence of DNA present in the list DNA\n",
    "def Motifs(Profile, Dna):\n",
    "    kmerList = []\n",
    "    for seq in Dna:\n",
    "        kmerList.append(ProfileMostProbablePattern(seq, len(Profile['A']), Profile))\n",
    "    return kmerList\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#RandomMotifs(Dna, k, t) uses \n",
    "#random.randint to choose a random k-mer from each of t different strings Dna, and returns a list of t strings.\n",
    "def RandomMotifs(Dna, k, t):\n",
    "    import random\n",
    "    finalKmer = []\n",
    "    for seq in Dna:\n",
    "        tempKmer = []\n",
    "        for i in range(len(seq)-k+1):\n",
    "            tempKmer.append(seq[i:i+k])\n",
    "        finalKmer.append(tempKmer[random.randint(0, len(tempKmer)-1)])\n",
    "    return finalKmer\n",
    "        \n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function will keep finding random motifs that are better than the previous one for N-runs\n",
    "def RandomizedMotifSearch(Dna, k, t, N):\n",
    "    M = RandomMotifs(Dna, k, t)\n",
    "    BestMotifs = M\n",
    "    runs = 0\n",
    "    #A list of random Motifs has been generated\n",
    "    while True & runs < N:\n",
    "        #Generate a profile based on the motifs\n",
    "        runs +=1\n",
    "        Profile = ProfileWithPseudocounts(M)\n",
    "        M = Motifs(Profile, Dna)\n",
    "        if Score(M) < Score(BestMotifs):\n",
    "            BestMotifs = M\n",
    "        else:\n",
    "            return BestMotifs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "##The following functions are meant to improve the randomized counting method "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function takes a set of probabilities and normalizes them so they add up to 1\n",
    "def Normalize(Probabilities):\n",
    "    sum = 0\n",
    "    normProb = {}\n",
    "    for key in Probabilities:\n",
    "        normProb[key] = 0\n",
    "        sum += Probabilities[key]\n",
    "    factor = 1.0/sum\n",
    "    for key in normProb:\n",
    "        normProb[key] = Probabilities[key] * factor\n",
    "    return normProb\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#return a random k-mer given a dictionary of k-mers and probabilities\n",
    "def WeightedDie(Probabilities):\n",
    "    import random\n",
    "    newProb = Normalize(Probabilities)\n",
    "    kmer = '' \n",
    "    dieRoll = random.uniform(0, 1)\n",
    "    iterableSum = 0\n",
    "    for seq in newProb:\n",
    "        iterableSum += newProb[seq]\n",
    "        if iterableSum >= dieRoll:\n",
    "            kmer = seq\n",
    "            break\n",
    "    return kmer\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ProfileGeneratedString(Text, profile, k):\n",
    "    import random\n",
    "    n = len(Text)\n",
    "    probabilities = {} \n",
    "    for i in range(0,n-k+1):\n",
    "        probabilities[Text[i:i+k]] = Pr(Text[i:i+k], profile)\n",
    "    return WeightedDie(probabilities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The final program\n",
    "def GibbsSampler(Dna, k, t, N):\n",
    "    import random\n",
    "    BestMotifs = []\n",
    "    M = RandomMotifs(Dna, k, t)\n",
    "    BestMotifs = M\n",
    "    #print(\"The first list of random k-mers:\" + str(M))\n",
    "    for j in range(N):\n",
    "        i = random.randint(0,t-1)\n",
    "        del M[i]\n",
    "        #print(\"The k-mers with one random one removed: \"+str(M))\n",
    "        \n",
    "        Profile = ProfileWithPseudocounts(M)\n",
    "        newKmer = ProfileGeneratedString(Dna[i], Profile, k)\n",
    "        M.insert(i,newKmer)\n",
    "        #print(\"The k-mer list with a better k-mer added back in: \" + str(M))\n",
    "        if Score(M) < Score(BestMotifs):\n",
    "            BestMotifs = M\n",
    "    return BestMotifs\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TGA', 'TAA', 'GGA', 'TGT']\n"
     ]
    }
   ],
   "source": [
    "Dna = ['TGACGTTC',\n",
    "'TAAGAGTT',\n",
    "'GGACGAAA',\n",
    "'CTGTTCGC']\n",
    "M = ['TGA',\n",
    "'GTT',\n",
    "'GAA',\n",
    "'TGT']\n",
    "print(Motifs(Profile(M), Dna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
